io/StacReader.cpp

```diff
@@ -0,0 +1,521 @@
+        if (!m_args->bounds.valid())
+            throw pdal_error("Supplied bounds are not valid.");
+        log()->get(LogLevel::Debug) << "Bounds: " << m_args->bounds << std::endl;
+    }
+
+    // array of pipeline-like reader definitions
+    //{ "type": "readers.ept" , "resolution": 100, "bounds": "([x,x],[y,y])"}
+    if (!m_args->readerArgs.empty())
+    {
+        for (auto& opts: m_args->readerArgs)
+            if (!opts.is_object())
+                throw pdal_error("Reader Args must be a valid JSON object");
+
+        handleReaderArgs();
+    }
+
+    if (!m_args->assetName.empty())
+        log()->get(LogLevel::Debug) << "STAC Reader will look for assets in "
+            "asset name '" << m_args->assetName << "'." << std::endl;
+
+    if (m_args->dryRun)
+        log()->get(LogLevel::Debug) << "Dry Run flag is set." << std::endl;
+
+    if (m_args->schemaValidate)
+        log()->get(LogLevel::Debug) <<
+            "JSON Schema validation flag is set." << std::endl;
+
+}
+
+void StacReader::initialize()
+{
+    initializeArgs();
+
+    m_arbiter.reset(new arbiter::Arbiter());
+    std::string stacStr = m_arbiter->get(m_filename);
+    NL::json stacJson = NL::json::parse(stacStr);
+
+    if (!stacJson.contains("type"))
+        throw pdal_error("Invalid STAC object provided.");
+
+    std::string stacType = stacJson["type"];
+    if (stacType == "Feature")
+        initializeItem(stacJson);
+    else if (stacType == "Catalog")
+        initializeCatalog(stacJson);
+    else
+        throw pdal_error("Could not initialize STAC object of type " + stacType);
+}
+
+void schemaFetch(const nlohmann::json_uri &json_uri, nlohmann::json &json)
+{
+    std::unique_ptr<arbiter::Arbiter> fetcher;
+    fetcher.reset(new arbiter::Arbiter());
+    std::string jsonStr = fetcher->get(json_uri.url());
+    json = nlohmann::json::parse(jsonStr);
+}
+
+void StacReader::schemaValidate(NL::json stacJson)
+{
+    std::function<void(const nlohmann::json_uri &, nlohmann::json &)> fetch = schemaFetch;
+    nlohmann::json_schema::json_validator val(
+        fetch,
+        [](const std::string &, const std::string &) {}
+    );
+    if (!stacJson.contains("type"))
+        throw pdal_error("Invalid STAC json");
+    std::string type = stacJson["type"].get<std::string>();
+    std::string schemaUrl;
+
+    if (type == "Feature")
+    {
+        schemaUrl = "https://schemas.stacspec.org/v1.0.0/item-spec/json-schema/item.json";
+        for (auto& extSchemaUrl: stacJson["stac_extensions"])
+        {
+            log()->get(LogLevel::Debug) << "Processing extension " << extSchemaUrl << std::endl;
+            std::string schemaStr = m_arbiter->get(extSchemaUrl);
+            NL::json schemaJson = NL::json::parse(schemaStr);
+            val.set_root_schema(schemaJson);
+            val.validate(stacJson);
+        }
+    }
+    else if (type == "Catalog")
+        schemaUrl = "https://schemas.stacspec.org/v1.0.0/catalog-spec/json-schema/catalog.json";
```

Comment:
@kylemann16 this URL should be a defaulted argument so the users can control which URL they want to try to apply schema validation too